### PR TITLE
Updated NodeHealthCheck with desired fields and add self node remediation api

### PR DESCRIPTION
**Pull request by geetikakay**
…

##### Short description:
1. Update NHC parameters
2. Add self node remediation api-resource in resource.py

### CODE REVIEW

File: ocp_resources/node_health_check.py [hunk 1]
```diff
@@ -8,3 +8,73 @@ class NodeHealthCheck(Resource):
     """
 
     api_group = Resource.ApiGroup.REMEDIATION_MEDIK8S_IO
+    self_node_remediation_api_group = Resource.ApiGroup.SELF_NODE_REMEDIATION_MEDIK8S_IO
+
+    def __init__(
+        self,
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:14]: - please use `kwargs` for anyting that is not related to the specific resource.
- The resource inherits from `Resource` yet a ns is added here

File: ocp_resources/node_health_check.py [hunk 2]
```diff
@@ -8,3 +8,73 @@ class NodeHealthCheck(Resource):
...
+
+        """
+        super().__init__(
+            name=name,
+            **kwargs,
+        )
+        self.node_role = node_role
+        self.min_unhealthy = min_unhealthy
+        self.unhealthy_duration = unhealthy_duration
+        self.escalating_remediation = escalating_remediation
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:45]: are they all optional?

File: ocp_resources/node_health_check.py [hunk 3]
```diff
@@ -8,3 +8,73 @@ class NodeHealthCheck(Resource):
...
+            if self.escalating_remediation:
+                self.res["spec"]["escalatingRemediations"] = self.escalating_remediation
+            if not self.escalating_remediation:
+                # Self Node Remediation is the default remediation for now.
+                self.res["spec"]["remediationTemplate"] = {
+                    "apiVersion": f"{self.self_node_remediation_api_group}/{self.api_version}.split('/')[1]",
+                    "kind": "SelfNodeRemediationTemplate",
+                    "name": "self-node-remediation-resource-deletion-template",
+                    "namespace": self.namespace,
+                }
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:60]: do not hard-code anything here; should be provided by the user

File: ocp_resources/node_health_check.py [hunk 4]
```diff
@@ -8,3 +8,73 @@ class NodeHealthCheck(Resource):
...
+        self.res["spec"]["minHealthy"] = self.min_unhealthy
+        self.res["spec"].setdefault("selector", {})
+        self.res["spec"]["selector"]["matchExpressions"] = [
+            {
+                "key": f"node-role.kubernetes.io/{self.node_role}",
+                "operator": "Exists",
+            }
+        ]
+        self.res["spec"]["unhealthyConditions"] = [
+            {
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:70]: do not set hard-coded values; you can provide defaults for this one

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/node_health_check.py (hunk 3 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py]: ack

File: ocp_resources/node_health_check.py (hunk 4 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py]: ack

File: ocp_resources/node_health_check.py (hunk 2 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py]: I have added in docstrings but escalating_remediation is mandatory if remediation_template don't exist

**Comment by geetikakay** [comment]
@rnetser done with changes and verified

File: ocp_resources/node_health_check.py (hunk 1 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py:13]: okay

File: ocp_resources/node_health_check.py [hunk 5]
```diff
@@ -8,3 +8,67 @@ class NodeHealthCheck(Resource):
...
+        min_unhealthy=None,
+        unhealthy_conditions=None,
+        escalating_remediation=None,
+        remediation_template=None,
+        yaml_file=None,
+        **kwargs,
+    ):
+        """
+           Args
+        node_role (str): role of the node. example: worker, control-plane
```
Comment (myakove) [inline ocp_resources/node_health_check.py:24]: missing `Optional` in `node_role`

**Review by myakove — CHANGES_REQUESTED** [review]

File: ocp_resources/node_health_check.py [hunk 6]
```diff
@@ -8,3 +8,67 @@ class NodeHealthCheck(Resource):
...
     api_group = Resource.ApiGroup.REMEDIATION_MEDIK8S_IO
+
+    def __init__(
+        self,
+        node_role="worker",
+        min_unhealthy=None,
+        unhealthy_conditions=None,
+        escalating_remediation=None,
+        remediation_template=None,
+        yaml_file=None,
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply ocp_resources/node_health_check.py:19]: ack

**Comment by geetikakay** [comment]
@myakove Hi, done with all the changes

**Review by myakove — DISMISSED** [review]

File: ocp_resources/node_health_check.py [hunk 7]
```diff
@@ -8,3 +8,66 @@ class NodeHealthCheck(Resource):
     """
 
     api_group = Resource.ApiGroup.REMEDIATION_MEDIK8S_IO
+
+    def __init__(
+        self,
+        node_role="worker",
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:14]: do not set a default here

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/node_health_check.py (hunk 7 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py:14]: after installing operator if i check yaml , default yaml has:

          "selector": {
              "matchExpressions": [
                {
                  "key": "node-role.kubernetes.io/worker",
                  "operator": "Exists"
                }
              ]
            },

File: ocp_resources/node_health_check.py [hunk 8]
```diff
@@ -8,3 +8,66 @@ class NodeHealthCheck(Resource):
...
+        **kwargs,
+    ):
+        """
+        Args
+        node_role (str, Optional): role of the node. example: worker, control-plane
+        min_unhealthy (int, Optional): Remediation is allowed if at least "MinHealthy" nodes selected by "selector"
+        Default value id 51%
+        unhealthy_conditions (list, Optional): UnhealthyConditions contains a list of the conditions that determine
+        whether a node is considered unhealthy.
+        remediation_template (dict, mandatory) : remediation template provided by an infrastructure provider.
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply ocp_resources/node_health_check.py:28]: ack i will do this change

File: ocp_resources/node_health_check.py [hunk 9]
```diff
@@ -8,3 +8,66 @@ class NodeHealthCheck(Resource):
...
+                raise ValueError(
+                    "Mandatory to have either RemediationTemplate or"
+                    " EscalatingRemediations"
+                )
+            if self.escalating_remediation and self.remediation_template:
+                raise ValueError(
+                    "RemediationTemplate and EscalatingRemediations usage is mutual"
+                    " exclusive"
+                )
+            self.res.setdefault("spec", {})
```
[the remark this answers is not in the document]
Comment (geetikakay) [reply ocp_resources/node_health_check.py:58]: ack

File: ocp_resources/node_health_check.py (hunk 7 above)
Comment (rnetser) [reply ocp_resources/node_health_check.py:14]: you should not hard code the resource defaults; what if this is changed?

File: ocp_resources/node_health_check.py [hunk 10]
```diff
@@ -8,3 +8,66 @@ class NodeHealthCheck(Resource):
...
+        unhealthy_conditions=None,
+        escalating_remediation=None,
+        remediation_template=None,
+        **kwargs,
+    ):
+        """
+        Args
+        node_role (str, Optional): role of the node. example: worker, control-plane
+        min_unhealthy (int, Optional): Remediation is allowed if at least "MinHealthy" nodes selected by "selector"
+        Default value id 51%
```
[the remark this answers is not in the document]
Comment (rnetser) [reply ocp_resources/node_health_check.py:25]: this is set by the resource, not by the code in ocp wrapper. if this is changed in the resource, this will no longer be true

File: ocp_resources/node_health_check.py (hunk 7 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py:14]: it is mandatory to have this field.

File: ocp_resources/node_health_check.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (geetikakay) [reply ocp_resources/node_health_check.py:25]: okay i will remove this line

File: ocp_resources/node_health_check.py (hunk 7 above)
Comment (geetikakay) [reply ocp_resources/node_health_check.py:14]: does that mean node_role="worker" should have node_role=None?

Comment (geetikakay) [reply ocp_resources/node_health_check.py:14]: @rnetser Hi, ^

**Review by geetikakay — COMMENTED** [review]
please have a look

File: ocp_resources/node_health_check.py (hunk 7 above)
Comment (rnetser) [reply ocp_resources/node_health_check.py:14]: yes

**Comment by geetikakay** [comment]
@myakove Hi, Can i see why tox failed

File: ocp_resources/node_health_check.py [hunk 11]
```diff
@@ -8,3 +8,75 @@ class NodeHealthCheck(Resource):
...
+        This is a map of {key,value} pairs.
+        min_unhealthy (int, Optional): Remediation is allowed if at least "MinHealthy" nodes selected by "selector"
+        unhealthy_conditions (list, Optional): UnhealthyConditions contains a list of the conditions that determine
+        whether a node is considered unhealthy.
+        remediation_template (dict, mandatory if `escalating_remediation` is not provided) : remediation template
+        provided by an infrastructure provider.
+        escalating_remediation (list, mandatory if `remediation_template` is not provided): EscalatingRemediations
+        contain a list of ordered remediation
+        templates with a timeout.
+        example: [{'remediationTemplate': {'apiVersion': 'self-node-remediation.medik8s.io/v1alpha1', 'kind':
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:36]: indent is needed

File: ocp_resources/node_health_check.py [hunk 12]
```diff
@@ -8,3 +8,75 @@ class NodeHealthCheck(Resource):
...
+        super().to_dict()
+        if not self.yaml_file:
+            if not (self.selector_match_expressions or self.selector_match_labels):
+                raise ValueError(
+                    "Mandatory to have Selector either selector_match_expressions or"
+                    " selector_match_labels"
+                )
+            if not (self.escalating_remediation or self.remediation_template):
+                raise ValueError(
+                    "Mandatory to have either RemediationTemplate or"
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:61]: replace field names with arg names

File: ocp_resources/node_health_check.py [hunk 13]
```diff
@@ -8,3 +8,75 @@ class NodeHealthCheck(Resource):
...
+            spec = self.res.setdefault("spec", {})
+            if self.escalating_remediation:
+                spec["escalatingRemediations"] = self.escalating_remediation
+            if self.remediation_template:
+                spec["remediationTemplate"] = self.remediation_template
+            if self.min_unhealthy:
+                spec["minHealthy"] = self.min_unhealthy
+            if self.unhealthy_conditions:
+                spec["unhealthyConditions"] = self.unhealthy_conditions
+            spec.setdefault("selector", {})
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:78]: can selector be an empy dict? if `selector_match_expressions` and `selector_match_labels` are not passed

**Review by rnetser — CHANGES_REQUESTED** [review]

File: ocp_resources/node_health_check.py [hunk 14]
```diff
@@ -8,3 +8,75 @@ class NodeHealthCheck(Resource):
...
+        selector_match_expressions=None,
+        selector_match_labels=None,
+        min_unhealthy=None,
+        unhealthy_conditions=None,
+        escalating_remediation=None,
+        remediation_template=None,
+        **kwargs,
+    ):
+        """
+        Args
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:23]: please format according to  https://google.github.io/styleguide/pyguide.html

**Review by rnetser — CHANGES_REQUESTED** [review]

**Comment by geetikakay** [comment]
@rnetser Hi Ruth, I have done the changes. Please have a look. thanks

**Comment by geetikakay** [comment]
@myakove @rnetser Hi, This PR got closed. I have added the asked changes. Could you please reopen this. I don't have permissions to reopen

File: ocp_resources/node_health_check.py [hunk 15]
```diff
@@ -8,3 +8,69 @@ class NodeHealthCheck(Resource):
...
+        escalating_remediation=None,
+        remediation_template=None,
+        **kwargs,
+    ):
+        """
+        Create NodeHealthCheck object.
+
+        Args:
+            selector_match_expressions (dict, Optional if `selector_match_labels` exist): Belongs to selector object.
+            This helps in matching nodes.
```
Comment (rnetser) [inline ocp_resources/node_health_check.py:27]: indentation is missing in several places in the docstring.
```
<arg name (type)>: ....
    ....
```

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by rnetser — APPROVED** [review]

**Review by rnetser — APPROVED** [review]
